README.md
# WPS Hide Login Lite
Based on the WPS Hide Login plugin for WordPress, but with all bullshit removed.

This plugin lets you easily and safely change the url of the login form page to anything you want. It doesn’t literally rename or change files in core, nor does it add rewrite rules. It simply intercepts page requests and works on any WordPress website. The wp-admin directory and wp-login.php page become inaccessible, so you should bookmark or remember the url. Deactivating this plugin brings your site back exactly to the state it was before.


CHANGES.md
## Changes
= 1.9.2 =
* Add action before redirect
* Fix redirect with wp-cli (Thanks @netson)

= 1.9.1 =
* Fix : by-pass security issue allowing an unauthenticated user to get login page by setting a random referer string via curl request.

= 1.9 =
* Fix : redirect ajax add_to_cart

= 1.8.8 =
* Fix : redirect_url (Thanks Don)

= 1.8.7 =
* Fix : remove redirect in doing cron

= 1.8.6 =
* Tested up to 5.8 (please see entry v1.8.4 for rants)

= 1.8.5 =
* Fix : Force refresh permalinks update option 'whl_page'

= 1.8.4 =
* Tested up to 5.7 (what a brilliant achievement. you created a new version just because you tested a piece of software? who taught you about release management and coding?)

= 1.8.3 =
* Fix : remove WP_Review (that was never in my lite version, but hey, good to see, that you're learning)

= 1.8.2 =
* Fix notice "Notice: Trying to get property 'href' of non-object" (whereever that was ...)

= 1.8.1 =
* Fix fatal error with vendor wp-dismissible-notices-handler and wp-review-me (nope, not applicable for the lite version)

= 1.8 =
* Fix multisite subdomain for website menu (Thanks Eric Celeste)

= 1.7 =
* Fix vulnerability (Thanks Sebastian Schmitt) : Posting "post_password" with arbitrary content to /wp-login.php reveals the normal wordpress login page. (You "fixed" it by checking for an additonal GET parameter?)

= 1.6.1 =
* Fix : loopback request site-health

= 1.6 =
* Tested up to 5.6
* Add compatibility with PHP8 (which I can't find anywhere in the actual code; maybe it's in the Composer stuiff, which I stripped, so useless).

= 1.5.7 =
* Fix: Text Domain Issue (which is totally useless, because the Lite version does not make use of translation, because the translation is fucked). 

= 1.5.6 =
* Fix: flush rewrite rules after install or update option
* Tested up to 5.4

= 1.5.5 =
* Add filter to redirect in cases where the user is already logged in.
* Fix: add rawurldecode for all $_SERVER['REQUEST_URI'] (Thanks @nintechnet)

1.5.4.2
* Fix: changed check of deprected function icl_object_id to wpml_object_id

1.5.4.1
* Fix: home_url / site_url

1.5.4
* Fix: Compatibility with WPML (Thanks @susansiow)

1.5.3
* Fix: Security vulnerabilities (Thanks @juliobox)

1.5.2.2
* Tested up to 5.2
* Fix: Domain language

1.5.2.1
* Fix: Notice: Undefined index: query
* Remove: 111KB of bullshit.

1.5.2
* Fix: Action URL wp_send_user_request()

1.5.1
* Fix: Action URL get_the_password_form()

1.5
* Enhancement: Add custom redirection URL

1.4.5
* Fix: function wp_login_url on page 404 now returns an empty link

1.4.4
* Fix: Too many redirects when a user clicks “Log in with WordPress.com”

1.4.3
* Fix: Fatal Error with multisite WP

1.4.2
* Fix: Error with library for compat WordPress and PHP

1.4.1
* Fix: Remove message review if PHP is too old

1.4
* Enhancement code with composer, namespace and autoload

1.3.4.2
* Fix: Remove message review if PHP is too old

1.3.4.1
* Fix: Deprecated method

1.3.4
* Add: Review message
* Fix: Redirect url wp-admin/options.php

1.3.3
* Add: Filter hook for enable wp-signup (@sumobi)

1.3.2
* Fix: Encoding of the login with a space in the emails

1.3.1
* Fix: redirect change admin email

1.3
* Fix: redirect wp-register.php

1.2.7
* Enhancement for Woocommerce email notification

1.2.6.1
* Revert redirect after login

1.2.6
* Fix: redirect after login

1.2.5.1
* Fix: add action in hook activate

1.2.5
* Remove: redirect activate

1.2.4
* Remove: Third party wpserveur

1.2.3.1
* Enhancement: Add translations cs_CZ, da_DK, es_ES, it_IT, ru_RU
* Fix: Parse error classes/plugin.php l.530

1.2.3
* Fix: change 403 to 404 error on wp-admin
* Fix: activate plugin
* Enhancement: Third party wpserveur

1.2.2
* Enhancement: Compatibility 4.9.x

1.2.1
* Enhancement: Prevent access to the login page by using the URL encoded version of wp-login.php

1.2
* Enhancement: Prevent redirection to login URL when accessing /wp-admin/customize.php directly
* Enhancement: Redirect to admin URL when already logged-in and accessing login URL without the action query string

1.1.7
* Fix: change fake 404 on wp-admin when not logged-in to a 403 forbidden to prevent fatal errors with various themes & plugins

1.1.6
* Fix: bug with Yoast SEO causing a Fatal Error and blank screen when loading /wp-admin/ without being logged-in

1.1.5
* Fix: Stop displaying the new login url notice everywhere when settings are updated (thanks @ kmelia on GitHub)
* Improvement: better way of retrieving the 404 template

1.1.4
* Fix: bypass the plugin when $pagenow is admin-post.php

1.1.3
* Fix: issue if no 404 template in active theme directory

1.1.2
* Modified priority on hooks to fix a problem with some configurations

1.1.1
* Check for Rename wp-login.php activation before activating WPS Hide Login Lite to prevent conflict

1.1
* Fix: CSRF security issue when saving option value in single site and multisite mode. Thanks to @Secupress

wps-hide-login/uninstall.php
<?php
/**
 * Fired when the plugin is uninstalled.
 *
 * @package   WPS Hide Login
 * @author    Remy Perona <remperona@gmail.com>
 * @license   GPL-2.0+
 * @link      http://remyperona.fr
 * @copyright 2015 Remy Perona
 */

// If uninstall not called from WordPress, then exit
if ( ! defined( 'WP_UNINSTALL_PLUGIN' ) ) {
	exit;
}

global $wpdb;

if ( is_multisite() ) {

	$blogs = $wpdb->get_results( "SELECT blog_id FROM {$wpdb->blogs}", ARRAY_A );
	delete_site_option( 'whl_page' );
	delete_site_option( 'whl_redirect_admin' );

	flush_rewrite_rules();
	
	if ( $blogs ) {

		foreach ( $blogs as $blog ) {
			switch_to_blog( $blog['blog_id'] );
			delete_option( 'whl_page' );
			delete_option( 'whl_redirect_admin' );
			
			flush_rewrite_rules();
			
			//info: optimize table
			$GLOBALS['wpdb']->query( "OPTIMIZE TABLE `" . $GLOBALS['wpdb']->prefix . "options`" );
			restore_current_blog();
		}
	}

} else {
	delete_option( 'whl_page' );
	delete_option( 'whl_redirect_admin' );

	flush_rewrite_rules();
	
	//info: optimize table
	$GLOBALS['wpdb']->query( "OPTIMIZE TABLE `" . $GLOBALS['wpdb']->prefix . "options`" );
}

wps-hide-login/wps-hide-login.php
<?php
/*
 Plugin Name: WPS Hide Login Lite
 Description: Protect your website by changing the login URL and preventing access to wp-login.php page and wp-admin directory while not logged-in. This lite version is based on the latest version of WPS Hide Login, but with all bullshit removed.
 Author: NicolasKulka, wpformation, nrekow
 Author URI: https://rekow.ch
 Version: 1.9.2
 Requires at least: 4.1
 Tested up to: 5.8
 Requires PHP: 7.0
 License: GPLv3
 License URI: http://www.gnu.org/licenses/gpl-3.0.html
 */

// Don't load directly
if (!defined('ABSPATH')) {
	die();
}


// Plugin constants
define( 'WPS_HIDE_LOGIN_VERSION', '1.9.2' );
define( 'WPS_HIDE_LOGIN_FOLDER', 'wps-hide-login' );

define( 'WPS_HIDE_LOGIN_URL', plugin_dir_url( __FILE__ ) );
define( 'WPS_HIDE_LOGIN_DIR', plugin_dir_path( __FILE__ ) );
define( 'WPS_HIDE_LOGIN_BASENAME', plugin_basename( __FILE__ ) );

// Function to easily include files.
function wps_hide_login_load_files( $dir, $files, $prefix = '') {
	foreach ($files as $file) {
		if (is_file($dir . $prefix . $file . '.php')) {
			require_once $dir . $prefix . $file . '.php';
		}
	}
}


// Plugin client classes
wps_hide_login_load_files( WPS_HIDE_LOGIN_DIR . 'classes/', array('singleton', 'plugin') );


// Activation hook
register_activation_hook( __FILE__, array( '\WPS\WPS_Hide_Login\Plugin', 'activate' ) );

function plugins_loaded_wps_hide_login_plugin() {
	\WPS\WPS_Hide_Login\Plugin::get_instance();
}
add_action( 'plugins_loaded', 'plugins_loaded_wps_hide_login_plugin' );


wps-hide-login/classes/plugin.php
<?php

namespace WPS\WPS_Hide_Login;

class Plugin {
	use Singleton;

	private $wp_login_php;
	
	protected function init() {
		global $wp_version;

		if ( version_compare( $wp_version, '4.0-RC1-src', '<' ) ) {
			add_action( 'admin_notices', array( $this, 'admin_notices_incompatible' ) );
			add_action( 'network_admin_notices', array( $this, 'admin_notices_incompatible' ) );

			return;
		}

		if ( is_multisite() && ! function_exists( 'is_plugin_active_for_network' ) || ! function_exists( 'is_plugin_active' ) ) {
			require_once( ABSPATH . '/wp-admin/includes/plugin.php' );
		}

		if ( is_plugin_active_for_network( 'rename-wp-login/rename-wp-login.php' ) ) {
			deactivate_plugins( WPS_HIDE_LOGIN_BASENAME );
			add_action( 'network_admin_notices', array( $this, 'admin_notices_plugin_conflict' ) );
			
			if ( isset( $_GET['activate'] ) ) {
				unset( $_GET['activate'] );
			}

			return;
		}

		if ( is_plugin_active( 'rename-wp-login/rename-wp-login.php' ) ) {
			deactivate_plugins( WPS_HIDE_LOGIN_BASENAME );
			add_action( 'admin_notices', array( $this, 'admin_notices_plugin_conflict' ) );
			
			if ( isset( $_GET['activate'] ) ) {
				unset( $_GET['activate'] );
			}

			return;
		}

		if ( is_multisite() && is_plugin_active_for_network( WPS_HIDE_LOGIN_BASENAME ) ) {
			add_action( 'wpmu_options', array( $this, 'wpmu_options' ) );
			add_action( 'update_wpmu_options', array( $this, 'update_wpmu_options' ) );
			add_filter( 'network_admin_plugin_action_links_' . WPS_HIDE_LOGIN_BASENAME, array(
				$this,
				'plugin_action_links'
			) );
		}

		if ( is_multisite() ) {
			add_action( 'wp_before_admin_bar_render', array( $this, 'modify_mysites_menu' ), 999 );
		}
		
		add_action( 'admin_init', array( $this, 'admin_init' ) );
		add_action( 'plugins_loaded', array( $this, 'plugins_loaded' ), 9999 );
		add_action( 'admin_notices', array( $this, 'admin_notices' ) );
		add_action( 'network_admin_notices', array( $this, 'admin_notices' ) );
		add_action( 'wp_loaded', array( $this, 'wp_loaded' ) );
		add_action( 'setup_theme', array( $this, 'setup_theme' ), 1 );

		add_filter( 'plugin_action_links_' . WPS_HIDE_LOGIN_BASENAME, array( $this, 'plugin_action_links' ) );
		add_filter( 'site_url', array( $this, 'site_url' ), 10, 4 );
		add_filter( 'network_site_url', array( $this, 'network_site_url' ), 10, 3 );
		add_filter( 'wp_redirect', array( $this, 'wp_redirect' ), 10, 2 );
		add_filter( 'site_option_welcome_email', array( $this, 'welcome_email' ) );

		remove_action( 'template_redirect', 'wp_redirect_admin_locations', 1000 );
		add_action( 'admin_enqueue_scripts', array( $this, 'admin_enqueue_scripts' ) );

		add_action( 'admin_menu', array( $this, 'wps_hide_login_menu_page' ) );
		add_action( 'admin_init', array( $this, 'whl_template_redirect' ) );

		add_action( 'template_redirect', array( $this, 'redirect_export_data' ) );
		add_filter( 'login_url', array( $this, 'login_url' ), 10, 3 );

		add_filter( 'user_request_action_email_content', array( $this, 'user_request_action_email_content' ), 999, 2 );
		
		add_filter( 'site_status_tests', array( $this, 'site_status_tests' ) );
	}
	
	public function site_status_tests( $tests ) {
	    unset( $tests['async']['loopback_requests'] );
	    
	    return $tests;
	}
	
	public function user_request_action_email_content( $email_text, $email_data ) {
		return str_replace( '###CONFIRM_URL###', esc_url_raw( str_replace( $this->new_login_slug() . '/', 'wp-login.php', $email_data['confirm_url'] ) ), $email_text );
	}

	
	private function use_trailing_slashes() {
		return ( '/' === substr( get_option( 'permalink_structure' ), - 1, 1 ) );
	}


	private function user_trailingslashit( $string ) {
		return $this->use_trailing_slashes() ? trailingslashit( $string ) : untrailingslashit( $string );
	}

	private function wp_template_loader() {
		global $pagenow;

		$pagenow = 'index.php';
		if ( ! defined( 'WP_USE_THEMES' ) ) {
			define( 'WP_USE_THEMES', true );
		}

		wp();
		require_once( ABSPATH . WPINC . '/template-loader.php' );
		die();
	}

	public function modify_mysites_menu() {
		global $wp_admin_bar;
		
		$all_toolbar_nodes = $wp_admin_bar->get_nodes();
		
		foreach ( $all_toolbar_nodes as $node ) {
			if ( preg_match( '/^blog-(\d+)(.*)/', $node->id, $matches ) ) {
				$blog_id = $matches[1];
				if ( $login_slug = $this->new_login_slug( $blog_id ) ) {
					if ( ! $matches[2] || '-d' === $matches[2] ) {
						$args       = $node;
						$old_href   = $args->href;
						$args->href = preg_replace( '/wp-admin\/$/', "$login_slug/", $old_href );
						if ( $old_href !== $args->href ) {
							$wp_admin_bar->add_node( $args );
						}
					} elseif ( strpos( $node->href, '/wp-admin/' ) !== false ) {
						$wp_admin_bar->remove_node( $node->id );
					}
				}
			}
		}
	}
	
	private function new_login_slug( $blog_id = '' ) {
		if ( $blog_id ) {
			if ( $slug = get_blog_option( $blog_id, 'whl_page' ) ) {
				return $slug;
			}
		} else {
			if ( $slug = get_option( 'whl_page' ) ) {
				return $slug;
			} else if ( ( is_multisite() && is_plugin_active_for_network( WPS_HIDE_LOGIN_BASENAME ) && ( $slug = get_site_option( 'whl_page', 'login' ) ) ) ) {
				return $slug;
			} else if ( $slug = 'login' ) {
				return $slug;
			}
		}
	}
	
	private function new_redirect_slug() {
		if ( $slug = get_option( 'whl_redirect_admin' ) ) {
			return $slug;
		} else if ( ( is_multisite() && is_plugin_active_for_network( WPS_HIDE_LOGIN_BASENAME ) && ( $slug = get_site_option( 'whl_redirect_admin', '404' ) ) ) ) {
			return $slug;
		} else if ( $slug = '404' ) {
			return $slug;
		}
	}

	public function new_login_url( $scheme = null ) {
		$url = home_url( '/', $scheme );
		if ( function_exists('wpml_object_id') ) {
			$url = site_url( '/', $scheme );
		}
		
		$url = apply_filters( 'wps_hide_login_home_url', $url );
		
		if ( get_option( 'permalink_structure' ) ) {
			return $this->user_trailingslashit( $url . $this->new_login_slug() );
		} else {
			return $url . '?' . $this->new_login_slug();
		}
	}

	public function new_redirect_url( $scheme = null ) {
		$url = home_url( '/', $scheme );
		if ( function_exists('wpml_object_id') ) {
			$url = site_url( '/', $scheme );
		}

		if ( get_option( 'permalink_structure' ) ) {
			return $this->user_trailingslashit( $url . $this->new_redirect_slug() );
		} else {
			return $url . '?' . $this->new_redirect_slug();
		}
	}

	public function admin_notices_incompatible() {
		echo '<div class="error notice is-dismissible"><p>' . __( 'Please upgrade to the latest version of WordPress to activate', 'wps-hide-login' ) . ' <strong>' . __( 'WPS Hide Login', 'wps-hide-login' ) . '</strong>.</p></div>';
	}

	public function admin_notices_plugin_conflict() {
		echo '<div class="error notice is-dismissible"><p>' . __( 'WPS Hide Login could not be activated because you already have Rename wp-login.php active. Please uninstall rename wp-login.php to use WPS Hide Login', 'wps-hide-login' ) . '</p></div>';
	}

	/**
	 * Plugin activation
	 */
	public static function activate() {
		//add_option( 'whl_redirect', '1' );
		do_action( 'wps_hide_login_activate' );
	}

	public function wpmu_options() {
		$out = '';
		$out .= '<h3>' . __( 'WPS Hide Login', 'wps-hide-login' ) . '</h3>';
		$out .= '<p>' . __( 'This option allows you to set a networkwide default, which can be overridden by individual sites. Simply go to to the site’s permalink settings to change the url.', 'wps-hide-login' ) . '</p>';
		$out .= '<table class="form-table">';
		$out .= '<tr valign="top">';
		$out .= '<th scope="row"><label for="whl_page">' . __( 'Networkwide default', 'wps-hide-login' ) . '</label></th>';
		$out .= '<td><input id="whl_page" type="text" name="whl_page" value="' . esc_attr( get_site_option( 'whl_page', 'login' ) ) . '"></td>';
		$out .= '<th scope="row"><label for="whl_redirect_admin">' . __( 'Redirection url default', 'wps-hide-login' ) . '</label></th>';
		$out .= '<td><input id="whl_redirect_admin" type="text" name="whl_redirect_admin" value="' . esc_attr( get_site_option( 'whl_redirect_admin', '404' ) ) . '"></td>';
		$out .= '</tr>';
		$out .= '</table>';
		echo $out;
	}

	public function update_wpmu_options() {
		if ( ! empty( $_POST ) && check_admin_referer( 'siteoptions' ) ) {
			if ( ( $whl_page = sanitize_title_with_dashes( $_POST['whl_page'] ) )
				&& strpos( $whl_page, 'wp-login' ) === false
				&& ! in_array( $whl_page, $this->forbidden_slugs() ) ) {

				flush_rewrite_rules(true);
				update_site_option( 'whl_page', $whl_page );
			}
			
			if ( ( $whl_redirect_admin = sanitize_title_with_dashes( $_POST['whl_redirect_admin'] ) )
				&& strpos( $whl_redirect_admin, '404' ) === false ) {

				flush_rewrite_rules(true);
				update_site_option( 'whl_redirect_admin', $whl_redirect_admin );
			}
		}
	}

	public function admin_init() {
		global $pagenow;

		add_settings_section(
			'wps-hide-login-section',
			'WPS Hide Login',
			array( $this, 'whl_section_desc' ),
			'general'
		);

		add_settings_field(
			'whl_page',
			'<label for="whl_page">' . __( 'Login url', 'wps-hide-login' ) . '</label>',
			array( $this, 'whl_page_input' ),
			'general',
			'wps-hide-login-section'
		);

		add_settings_field(
			'whl_redirect_admin',
			'<label for="whl_redirect_admin">' . __( 'Redirection url', 'wps-hide-login' ) . '</label>',
			array( $this, 'whl_redirect_admin_input' ),
			'general',
			'wps-hide-login-section'
		);

		register_setting( 'general', 'whl_page', 'sanitize_title_with_dashes' );
		register_setting( 'general', 'whl_redirect_admin', 'sanitize_title_with_dashes' );

		if ( get_option( 'whl_redirect' ) ) {

			delete_option( 'whl_redirect' );

			if ( is_multisite()
			     && is_super_admin()
			     && is_plugin_active_for_network( WPS_HIDE_LOGIN_BASENAME ) ) {

				$redirect = network_admin_url( 'settings.php#whl_settings' );

			} else {

				$redirect = admin_url( 'options-general.php#whl_settings' );

			}

			wp_safe_redirect( $redirect );
			die();
		}
	}

	public function whl_section_desc() {
		if (!is_multisite()) {
			echo '<p>' . __( 'This option allows you to set a custom login and redirection URL.', 'wps-hide-login' ) . '</p>';
		}
		
		if ( is_multisite()
		     && is_super_admin()
		     && is_plugin_active_for_network( WPS_HIDE_LOGIN_BASENAME ) ) {

			echo '<p>' . sprintf( __( 'To set a networkwide default, go to <a href="%s">Network Settings</a>.', 'wps-hide-login' ), network_admin_url( 'settings.php#whl_settings' ) ) . '</p>';
		}
	}

	public function whl_page_input() {
		$url = home_url();
		if ( function_exists('wpml_object_id') ) {
			$url = site_url();
		}

		if ( get_option( 'permalink_structure' ) ) {
			echo '<code>' . trailingslashit( $url ) . '</code> <input id="whl_page" type="text" name="whl_page" value="' . $this->new_login_slug() . '">' . ( $this->use_trailing_slashes() ? ' <code>/</code>' : '' );
		} else {
			echo '<code>' . trailingslashit( $url ) . '?</code> <input id="whl_page" type="text" name="whl_page" value="' . $this->new_login_slug() . '">';
		}

		echo '<p class="description">' . __( 'Protect your website by changing the login URL and preventing access to the wp-login.php page and the wp-admin directory to non-connected people.', 'wps-hide-login' ) . '</p>';
	}

	public function whl_redirect_admin_input() {
		$url = home_url();
		if ( function_exists('wpml_object_id') ) {
			$url = site_url();
		}

		if ( get_option( 'permalink_structure' ) ) {
			echo '<code>' . trailingslashit( $url ) . '</code> <input id="whl_redirect_admin" type="text" name="whl_redirect_admin" value="' . $this->new_redirect_slug() . '">' . ( $this->use_trailing_slashes() ? ' <code>/</code>' : '' );
		} else {
			echo '<code>' . trailingslashit( $url ) . '?</code> <input id="whl_redirect_admin" type="text" name="whl_redirect_admin" value="' . $this->new_redirect_slug() . '">';
		}

		echo '<p class="description">' . __( 'Redirect URL when someone tries to access the wp-login.php page and the wp-admin directory while not logged in.', 'wps-hide-login' ) . '</p>';
	}

	public function admin_notices() {
		global $pagenow;

		if ( ! is_network_admin()
		     && $pagenow === 'options-general.php'
		     && isset( $_GET['settings-updated'] )
		     && ! isset( $_GET['page'] ) ) {

			echo '<div class="updated notice is-dismissible"><p>' . sprintf( __( 'Your login page is now here: <strong><a href="%1$s">%2$s</a></strong>. Bookmark this page!', 'wps-hide-login' ), $this->new_login_url(), $this->new_login_url() ) . '</p></div>';
		}
	}

	public function plugin_action_links( $links ) {
		if ( is_network_admin()
		     && is_plugin_active_for_network( WPS_HIDE_LOGIN_BASENAME ) ) {
			array_unshift( $links, '<a href="' . network_admin_url( 'settings.php#whl_settings' ) . '">' . __( 'Settings', 'wps-hide-login' ) . '</a>' );
		} elseif ( ! is_network_admin() ) {
			array_unshift( $links, '<a href="' . admin_url( 'options-general.php#whl_settings' ) . '">' . __( 'Settings', 'wps-hide-login' ) . '</a>' );
		}

		return $links;
	}

	public function redirect_export_data() {
		if ( ! empty( $_GET ) && isset( $_GET['action'] ) && 'confirmaction' === $_GET['action'] && isset( $_GET['request_id'] ) && isset( $_GET['confirm_key'] ) ) {
			$request_id = (int) $_GET['request_id'];
			$key        = sanitize_text_field( wp_unslash( $_GET['confirm_key'] ) );
			$result     = wp_validate_user_request_key( $request_id, $key );
			
			if ( ! is_wp_error( $result ) ) {
				wp_redirect(
					add_query_arg(
						array(
							'action'      => 'confirmaction',
							'request_id'  => $_GET['request_id'],
							'confirm_key' => $_GET['confirm_key']
						), $this->new_login_url()
					)
				);
				exit();
			}
		}
	}

	public function plugins_loaded() {
		global $pagenow;

		if ( ! is_multisite()
		&& ( strpos( rawurldecode( $_SERVER['REQUEST_URI'] ) , 'wp-signup' ) !== false || strpos( rawurldecode( $_SERVER['REQUEST_URI'] ), 'wp-activate' ) !== false ) && apply_filters( 'wps_hide_login_signup_enable', false ) === false ) {
			wp_die( __( 'This feature is not enabled.', 'wps-hide-login' ) );
		}
						
		$request = parse_url( rawurldecode( $_SERVER['REQUEST_URI'] ) );
						
		if ( ( strpos( rawurldecode( $_SERVER['REQUEST_URI'] ), 'wp-login.php' ) !== false
		       || ( isset( $request['path'] ) && untrailingslashit( $request['path'] ) === site_url( 'wp-login', 'relative' ) ) )
		     && ! is_admin() ) {
			$this->wp_login_php = true;
			$_SERVER['REQUEST_URI'] = $this->user_trailingslashit( '/' . str_repeat( '-/', 10 ) );
			$pagenow = 'index.php';
		} elseif ( ( isset( $request['path'] ) && untrailingslashit( $request['path'] ) === site_url( $this->new_login_slug(), 'relative' ) )
		           || ( ! get_option( 'permalink_structure' )
		                && isset( $_GET[ $this->new_login_slug() ] )
		                && empty( $_GET[ $this->new_login_slug() ] ) ) ) {
			$pagenow = 'wp-login.php';
		} elseif ( ( strpos( rawurldecode( $_SERVER['REQUEST_URI'] ), 'wp-register.php' ) !== false
		             || ( isset( $request['path'] ) && untrailingslashit( $request['path'] ) === site_url( 'wp-register', 'relative' ) ) )
		           && ! is_admin() ) {
			$this->wp_login_php = true;
			$_SERVER['REQUEST_URI'] = $this->user_trailingslashit( '/' . str_repeat( '-/', 10 ) );
			$pagenow = 'index.php';
		}
	}

	public function setup_theme() {
		global $pagenow;

		if ( ! is_user_logged_in() && 'customize.php' === $pagenow ) {
			wp_die( __( 'This has been disabled', 'wps-hide-login' ), 403 );
		}
	}

	public function wp_loaded() {
		global $pagenow;

		$request = parse_url( rawurldecode( $_SERVER['REQUEST_URI'] ) );

		do_action( 'wps_hide_login_before_redirect', $request );
		
		if ( ! ( isset( $_GET['action'] ) && $_GET['action'] === 'postpass' && isset( $_POST['post_password'] ) ) ) {
			if ( is_admin() && ! is_user_logged_in() && ! defined( 'WP_CLI' ) && ! defined( 'DOING_AJAX' ) && ! defined( 'DOING_CRON' ) && $pagenow !== 'admin-post.php' && $request['path'] !== '/wp-admin/options.php' ) {
				wp_safe_redirect( $this->new_redirect_url() );
				die();
			}

			if ( ! is_user_logged_in() && isset( $_GET['wc-ajax'] ) && $pagenow === 'profile.php' ) {
				wp_safe_redirect( $this->new_redirect_url() );
				die();
			}
			
			if ( ! is_user_logged_in() && $request['path'] === '/wp-admin/options.php' ) {
				header('Location: ' . $this->new_redirect_url() );
				die;
			}
			
			if ( $pagenow === 'wp-login.php'
			     && $request['path'] !== $this->user_trailingslashit( $request['path'] )
			     && get_option( 'permalink_structure' ) ) {
				wp_safe_redirect( $this->user_trailingslashit( $this->new_login_url() )
				                  . ( ! empty( $_SERVER['QUERY_STRING'] ) ? '?' . $_SERVER['QUERY_STRING'] : '' ) );
				die();
			} elseif ( $this->wp_login_php ) {

				if ( ( $referer = wp_get_referer() )
				     && strpos( $referer, 'wp-activate.php' ) !== false
				     && ( $referer = parse_url( $referer ) )
				     && ! empty( $referer['query'] ) ) {

					parse_str( $referer['query'], $referer );

					@require_once WPINC . '/ms-functions.php';

					if ( ! empty( $referer['key'] )
					     && ( $result = wpmu_activate_signup( $referer['key'] ) )
					     && is_wp_error( $result )
					     && ( $result->get_error_code() === 'already_active'
					          || $result->get_error_code() === 'blog_taken' ) ) {

						wp_safe_redirect( $this->new_login_url()
						                  . ( ! empty( $_SERVER['QUERY_STRING'] ) ? '?' . $_SERVER['QUERY_STRING'] : '' ) );
						die();
					}
				}
				
				// nrekow: disabled for testing purposes.
				//$this->wp_template_loader();
				die();
			} elseif ( $pagenow === 'wp-login.php' ) {
				global $error, $interim_login, $action, $user_login;

				$redirect_to = admin_url();
				
				$requested_redirect_to = '';
				if ( isset( $_REQUEST['redirect_to'] ) ) {
					$requested_redirect_to = $_REQUEST['redirect_to'];
				}
				
				if ( is_user_logged_in() ) {
					$user = wp_get_current_user();
					if ( ! isset( $_REQUEST['action'] ) ) {
						$logged_in_redirect = apply_filters( 'whl_logged_in_redirect', $redirect_to, $requested_redirect_to, $user );
						wp_safe_redirect( $logged_in_redirect );
						die();
					}
				}
				
				@require_once ABSPATH . 'wp-login.php';
				die();
			}
		}
	}

	public function site_url( $url, $path, $scheme, $blog_id ) {
		return $this->filter_wp_login_php( $url, $scheme );
	}

	public function network_site_url( $url, $path, $scheme ) {
		return $this->filter_wp_login_php( $url, $scheme );
	}

	public function wp_redirect( $location, $status ) {
		if ( strpos( $location, 'https://wordpress.com/wp-login.php' ) !== false ) {
			return $location;
		}

		return $this->filter_wp_login_php( $location );
	}

	public function filter_wp_login_php( $url, $scheme = null ) {
		if ( strpos( $url, 'wp-login.php?action=postpass' ) !== false ) {
			return $url;
		}

		if ( strpos( $url, 'wp-login.php' ) !== false && strpos( wp_get_referer(), 'wp-login.php' ) === false ) {
			if ( is_ssl() ) {
				$scheme = 'https';
			}

			$args = explode( '?', $url );
			if ( isset( $args[1] ) ) {
				parse_str( $args[1], $args );

				if ( isset( $args['login'] ) ) {
					$args['login'] = rawurlencode( $args['login'] );
				}

				$url = add_query_arg( $args, $this->new_login_url( $scheme ) );
			} else {
				$url = $this->new_login_url( $scheme );
			}
		}

		return $url;
	}

	public function welcome_email( $value ) {
		return $value = str_replace( 'wp-login.php', trailingslashit( get_site_option( 'whl_page', 'login' ) ), $value );
	}

	public function forbidden_slugs() {
		$wp = new \WP;
		return array_merge( $wp->public_query_vars, $wp->private_query_vars );
	}

	/**
	 * Load scripts
	 */
	public function admin_enqueue_scripts( $hook ) {
		if ( 'options-general.php' != $hook ) {
			return false;
		}

		wp_enqueue_style( 'plugin-install' );

		wp_enqueue_script( 'plugin-install' );
		wp_enqueue_script( 'updates' );
		add_thickbox();
	}

	public function wps_hide_login_menu_page() {
		$title = __( 'WPS Hide Login' );

		add_options_page( $title, $title, 'manage_options', 'whl_settings', array(
			$this,
			'settings_page'
		) );
	}

	public function settings_page() {
		_e( 'WPS Hide Login' );
	}

	public function whl_template_redirect() {
		if ( ! empty( $_GET ) && isset( $_GET['page'] ) && 'whl_settings' === $_GET['page'] ) {
			wp_redirect( admin_url( 'options-general.php#whl_settings' ) );
			exit();
		}
	}

	/**
	 *
	 * Update url redirect : wp-admin/options.php
	 *
	 * @param $login_url
	 * @param $redirect
	 * @param $force_reauth
	 *
	 * @return string
	 */
	public function login_url( $login_url, $redirect, $force_reauth ) {
		if ( is_404() ) {
			return '#';
		}

		if ( $force_reauth === false ) {
			return $login_url;
		}

		if ( empty( $redirect ) ) {
			return $login_url;
		}

		$redirect = explode( '?', $redirect );

		if ( $redirect[0] === admin_url( 'options.php' ) ) {
			$login_url = admin_url();
		}

		return $login_url;
	}
}